### DISCUSSION TITLE
Nodes not recovering from error state

### DISCUSSION THREAD

**Discussion by alexa22013** [emqx/emqx#11194] [Q&A]
I installed a sample plugin via emqx dashboard, due to some error the plugin was not enabling and nodes went to error state. I restarted manully to recover from it, but even after restarting it coninues to be in previous error state

`2023-07-04T19:17:57.756586+00:00 [error] crasher: initial call: application_master:init/4, pid: <0.2527.0>, registered_name: [], exit: {{bad_return,{{emqx_plugins_app,start,[normal,[]]},{'EXIT',{#{errors => [{"sample_plugin-1.0.0",#{error => "bad_info_file",path => "plugins/sample_plugin-1.0.0/release.json",return => {enoent,"/opt/emqx/plugins/sample_plugin-1.0.0/release.json"}}}],function => fun emqx_plugins:do_ensure_started/1},[{emqx_plugins,for_plugins,1,[{file,"emqx_plugins.erl"},{line,680}]},{emqx_plugins_app,start,2,[{file,"emqx_plugins_app.erl"},{line,28}]},{application_master,start_it_old,4,[{file,"application_master.erl"},{line,293}]}]}}}},[{application_master,init,4,[{file,"application_master.erl"},{line,142}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}, ancestors: [<0.2526.0>], message_queue_len: 1, messages: [{'EXIT',<0.2528.0>,normal}], links: [<0.2526.0>,<0.1707.0>], dictionary: [], trap_exit: true, status: running, heap_size: 987, stack_size: 29, reductions: 210; neighbours:
`

How do I resolve this? the nodes are continously restarting..
Thanks advance

**Comment by lafirest** [comment]
@alexa22013 
The root cause is a description file missing, maybe you used the wrong path or deleted this file by mistake.
Here, you can clean up the data directory of EMQX when stopping, and restart, it would work.

**Reply by alexa22013** [reply]
Thanks! @lafirest for the solution. This happens whenever there is a plugin active and cluster is updated. On restart the pods go into continous restart and fails to start. Any way to prevent it?

**Reply by lafirest** [reply]
@alexa22013 
Sorry for this late reply, is there any log for your description?

**Reply by alexa22013** [reply]
pod logs:
```
2023-07-04T19:17:57.756586+00:00 [error] crasher: initial call: application_master:init/4, pid: <0.2527.0>, registered_name: [], exit: {{bad_return,{{emqx_plugins_app,start,[normal,[]]},{'EXIT',{#{errors => [{"sample_plugin-1.0.0",#{error => "bad_info_file",path => "plugins/sample_plugin-1.0.0/release.json",return => {enoent,"/opt/emqx/plugins/sample_plugin-1.0.0/release.json"}}}],function => fun emqx_plugins:do_ensure_started/1},[{emqx_plugins,for_plugins,1,[{file,"emqx_plugins.erl"},{line,680}]},{emqx_plugins_app,start,2,[{file,"emqx_plugins_app.erl"},{line,28}]},{application_master,start_it_old,4,[{file,"application_master.erl"},{line,293}]}]}}}},[{application_master,init,4,[{file,"application_master.erl"},{line,142}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,226}]}]}, ancestors: [<0.2526.0>], message_queue_len: 1, messages: [{'EXIT',<0.2528.0>,normal}], links: [<0.2526.0>,<0.1707.0>], dictionary: [], trap_exit: true, status: running, heap_size: 987, stack_size: 29, reductions: 210; neighbours: 
```

**Reply by thalesmg** [reply]
Hi @alexa22013 .

Which EMQX version are you using?

Did you use a persistent volume for the plugins directory?  If so, could you try manually removing the plugin directory from there?

**Reply by alexa22013** [reply]
> Which EMQX version are you using?

EMQX v5.0.14, operator 2.1.4, api v2alpha1
> Did you use a persistent volume for the plugins directory? If so, could you try manually removing the plugin directory from there?

the core nodes are mounted with persistent volume, i guess its mounted on path /opt/emqx/data not on /opt/emqx/plugins. So persistent volume are not used for plugins

**Reply by thalesmg** [reply]
I'll try to reproduce locally and see if there's a workaround.

Meanwhile, @Rory-Z would you have any hints for this case?  Perhaps manually editing config maps saved by operator?

**Reply by thalesmg** [reply]
@alexa22013 Do you have access to the dashboard UI?  Are there core nodes still running which you could access via `kubectl exec`?

**Reply by Rory-Z** [reply]
@thalesmg Sorry, I have no idea. In EMQX Operator 2.1 and EMQX 5.0, EMQX Operator didn't manager configMap for EMQX, it's just by EMQX self.

**Reply by thalesmg** [reply]
@alexa22013  If you can log into some already running container, you may try to manually remove the plugin with the following commands that need to be executed on a running node:

```sh
emqx eval-erl 'Name = <<"plugin_name-1.0.0">>, erpc:multicall(mria_mnesia:running_nodes(), fun() -> ok = emqx_plugins:ensure_stopped(Name), ok = emqx_plugins:ensure_disabled(Name), ok = emqx_plugins:ensure_uninstalled(Name) end).'

emqx eval-erl 'FileName = <<"plugin_name-1.0.0.tar.gz">>, erpc:multicall(mria_mnesia:running_nodes(), fun() -> ok = case file:delete(filename:join([emqx_plugins:install_dir(), FileName])) of ok -> ok; {error, enoent} -> ok end end).'
```

Replace `plugin_name-1.0.0` by your plugin name and version.

**Reply by alexa22013** [reply]
> Do you have access to the dashboard UI? Are there core nodes still running which you could access via kubectl exec?

in such case only one core node remains up, rest of the nodes remains in crash loop back
